enhancements/sig-architecture/8-addon-framework/README.md

```diff
@@ -0,0 +1,200 @@
+
+## Summary
+
+The is proposal is for a start to standardize the addon framework. The framework consists of two parts:
+- A lightweight part that only provide cluster registration and manifest resource distribution functions (registration-operator)
+- A bunch of addons that can easily enabled/disabled on the lightweight part.
+
+Commonly an addon will have an agent component that runs on managed cluster, which should have at least one of the following characteristic:
+
+1. It needs to talk to hub with certain ACLs
+2. Its configuration diffs in different managed cluster.
+
+we have `ClusterManagerAddon` and `ManagedClusterAddon` APIs defined already, so that an addon controller on the hub can be represented as a ClusterManagerAddon resource and an addon agent can be represented as a ManagedClusterAddon resource in  the cluster namespace on the hub.
+
+## Motivation
+
+With a standardize addon framework, it allows users to easily managed each function a separate addon. In addition, it will be possible to provide a an addon framework library so developers could be able to develop their own addon functions more easily.
+
+### Goals
+
+- Standardize the addon registration/enablement procedure
+    - Enable/disable addon
+    - Communication pattern
+    - Authn/Authz
+- Provide an addon framework interface to ease the addon development
+
+
+### Non-Goals
+
+- Implementation details of addon
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an open-cluster-management admin on hub, I can add an addon into the hub. For example, I want to install the submariner addon into the hub.
+
+#### Story 2
+
+As an open-cluster-management admin on hub, I can enable the addon installation on all managedclusters.
+
+#### Story 3
+
+As an admin on a managed cluster, I can enable and config the addon agent on a certain managed cluster. I was able to enable the addon agent on hub, which triggers the agent installation on the managed cluster.
+
+#### Story 4
+
+As an addon developer, I can easily develop the addon using the library without deep understanding on how agent manifests are distributed and registerd to hub.
+
+
+### Standardize addon procedure
+
+#### Register an addon on hub
+
+A common process for an addon to be installed on the hub cluster follows the procedure below:
+
+1. Cluster admin on hub deploy a certain addon controller
+2. Addon controller registers itself to the hub by creating a ClusterManagerAddon resource on hub
+```yaml
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ClusterManagementAddon
+metadata:
+ name: submariner
+spec:
+ addonMeta:
+   displayName: submariner
+   description: submariner-addon
+ addonConfiguration:
+   crd: submarineraddonconfig.addon.open-cluster-management.io
+   name: submariner-config
+```
+
+#### Enable addon agent on a managed cluster
+
+1. A user creates a ManagedClusterAddon resource of which the name is the same as the name of ClusterManagerAddon resource in the cluster namespace.
```

Comment:
This name mapping is very clear (this is great), but you should clearly indicate higher up that add-ons can only be added to a managedcluster one time.  It seem like a sensible restriction for a thing claiming to be an addon.

Reply:
updated

Reply:
Would rather these are governed by the `ClusterManagementAddon` controller injecting Addons (through templates) rather than a per-cluster configuration by the user.

Reply:
I think addon controller on hub can create `managedclusteraddon` cr based on corresponding `ClusterManagementAddon `. We could have interface on the addon framework library to specify how the addon controller want to deploy agent in multiple clusters (deploy in every cluster or honor a placement api). I will add a paragraph to clarify it.